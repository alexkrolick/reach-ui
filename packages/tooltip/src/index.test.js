import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import * as autoId from "@reach/auto-id"; // eslint-disable-line no-unused-vars
import Tooltip, { LEAVE_TIMEOUT, MOUSE_REST_TIMEOUT } from ".";

jest.mock("@reach/auto-id", () => ({
  ...jest.requireActual("@reach/auto-id")
}));

jest.mock("@reach/utils", () => ({
  ...jest.requireActual("@reach/utils"),
  checkStyles: jest.fn()
}));

beforeEach(() => {
  const newAutoId = jest.requireActual("@reach/auto-id");
  jest.setMock("@reach/auto-id", newAutoId);
  jest.useFakeTimers();
});

describe("Tooltip", () => {
  it("shows/hides on hover", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );
    const trigger = getByText("Trigger");

    expect(baseElement).toMatchSnapshot("closed tooltip");
    fireEvent.mouseOver(trigger);
    act(() => jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after mouse enter");
    fireEvent.mouseLeave(trigger);
    act(() => jest.advanceTimersByTime(LEAVE_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after mouse leave");
  });

  it("shows/hides when trigger is activeElement", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );
    const trigger = getByText("Trigger");

    expect(baseElement).toMatchSnapshot("before focus");
    fireEvent.focus(trigger);
    expect(baseElement).toMatchSnapshot("after focus");
    fireEvent.blur(trigger);
    act(() => jest.advanceTimersByTime(LEAVE_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after blur");
  });

  it("shows without timeout when one tooltip is already visible", () => {
    const { baseElement, getByText } = render(
      <>
        <Tooltip label="First">
          <button>First Trigger</button>
        </Tooltip>
        <Tooltip label="Second">
          <button>Second Trigger</button>
        </Tooltip>
      </>
    );
    const firstTrigger = getByText("First Trigger");
    const secondTrigger = getByText("Second Trigger");

    fireEvent.mouseOver(firstTrigger);
    act(() => jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after first mouseover");
    fireEvent.mouseLeave(firstTrigger);
    fireEvent.mouseOver(secondTrigger);
    expect(baseElement).toMatchSnapshot("after second mouseover");
  });

  it("hides on ESC", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );

    const trigger = getByText("Trigger");
    fireEvent.focus(trigger);
    act(() => jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot("after focus");
    fireEvent.keyDown(trigger, { key: "Escape" });
    expect(baseElement).toMatchSnapshot("after escape");
  });
});
