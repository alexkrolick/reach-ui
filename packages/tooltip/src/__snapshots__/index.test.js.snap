// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip hides on ESC: after escape 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip hides on ESC: after focus 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-5"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-5"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows without timeout when one tooltip is already visible: after first mouseover 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-3"
      data-reach-tooltip-trigger=""
    >
      First Trigger
    </button>
    <button
      data-reach-tooltip-trigger=""
    >
      Second Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-3"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      First
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows without timeout when one tooltip is already visible: after second mouseover 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      First Trigger
    </button>
    <button
      aria-describedby="tooltip-4"
      data-reach-tooltip-trigger=""
    >
      Second Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-4"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Second
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows/hides on hover: after mouse enter 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-1"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-1"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows/hides on hover: after mouse leave 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows/hides on hover: closed tooltip 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows/hides when trigger is activeElement: after blur 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows/hides when trigger is activeElement: after focus 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-2"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-2"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows/hides when trigger is activeElement: before focus 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;
